### PR TITLE
Topic/latest by

### PR DESCRIPTION
**Pull request by FrontierOtr**

### CODE REVIEW

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducer.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  var keyValue: KeyValue[String, ThriftFactValue] = new KeyValue[String, ThriftFactValue]
+  
+  def clear(): Unit = {
+    keyValue = new KeyValue[String, ThriftFactValue]
+  }
+
+  def update(fv: Fact): Unit = {
+    if (!fv.isTombstone) {
+      val factKey = fv.toThrift.getValue.getStructSparse.getV.get(key)
+      if (factKey != null && factKey.getS != null) {
```
Comment (charleso) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducer.scala]: There's almost no point checking `factKey.getS` here. It's either going to be set, or it's going to throw a runtime exception. You can call `isSetS` if you want.

**Comment by charleso** [comment]
Be good to have a CLI test for this. :( We need JSON ingest support, or a pre-ingest job to help.

File: ivory-core/src/test/scala/com/ambiata/ivory/core/gen/GenDictionary.scala [hunk 2]
```diff
@@ -100,12 +100,19 @@ object GenDictionary {
...
+            Gen.oneOf(
+              StructExpression(name, se),
+              if (Encoding.isNumeric(Expression.expressionEncoding(StructExpression(name, se), cd.encoding)))
+                Inverse(StructExpression(name, se))
+              else
+                Inverse(StructExpression(name, NumFlips))
+            )
           )
+        val subexpGen = Gen.oneOf(values.toList).flatMap {
+          case (name, StructEncodedValue(StringEncoding, false)) => Gen.frequency(5 -> subexpsWithInverses(name, StringEncoding, cd), 1 -> Gen.const(LatestBy(name)))
```
Comment (charleso) [inline ivory-core/src/test/scala/com/ambiata/ivory/core/gen/GenDictionary.scala]: Why the `false` here?

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducer.scala [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package com.ambiata.ivory.operation.extraction.reduction
+
+import com.ambiata.ivory.core.Fact
+import com.ambiata.ivory.core.thrift._
+import scala.collection.JavaConverters._ 
+
+case class LatestByReducer(key: String) extends Reduction {
+
+  var keyValue: KeyValue[String, ThriftFactValue] = new KeyValue[String, ThriftFactValue]
```
Comment (charleso) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducer.scala]: I would drop the type declaration here /supernit

File: ivory-core/src/test/scala/com/ambiata/ivory/core/gen/GenDictionary.scala (hunk 2 above)
Comment (FrontierOtr) [inline ivory-core/src/test/scala/com/ambiata/ivory/core/gen/GenDictionary.scala]: What would the LatestBy be for a feature where the 'by' field is optional?

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducer.scala (hunk 1 above)
Comment (FrontierOtr) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducer.scala]: Indeed, wrong check.

File: ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+      val newFact = fact.withValue(StructValue((others ++ List(asPrim)).toMap))
+      r.update(newFact)
+    })
+     
+    Value.fromThrift(r.save) ==== ListValue(keyVals.distinct.map(keyVal => {
+      val asPrim: (String, PrimitiveValue) = keyname -> StringValue(keyVal)
+      StructValue((others ++ List(asPrim)).toMap)
+    }))
+
+    r.clear()
```
Comment (charleso) [inline ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducerSpec.scala]: No point having this here - probably should be before the `foreach`.

File: ivory-core/src/test/scala/com/ambiata/ivory/core/gen/GenDictionary.scala (hunk 2 above)
Comment (charleso) [inline ivory-core/src/test/scala/com/ambiata/ivory/core/gen/GenDictionary.scala]: > What would the LatestBy be for a feature where the 'by' field is optional?

Sorry, got my meaning of `optional` reversed. Not sure if you want to check the optional flag in validation?

**Comment by charleso** [comment]
:+1:

File: ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  def latestBy = prop((fact: Fact, keyname: String, keyVals: List[String], others: List[(String, PrimitiveValue)]) => {
+    val r = new LatestByReducer(keyname)
+    keyVals.foreach(keyVal => {
+      val asPrim: (String, PrimitiveValue) = keyname -> StringValue(keyVal)
+      val newFact = fact.withValue(StructValue((others ++ List(asPrim)).toMap))
+      r.update(newFact)
+    })
+     
+    Value.fromThrift(r.save) ==== ListValue(keyVals.distinct.map(keyVal => {
+      val asPrim: (String, PrimitiveValue) = keyname -> StringValue(keyVal)
```
Comment (charleso) [inline ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/reduction/LatestByReducerSpec.scala]: Not sure I would bother with having the `asPrim` values. /supernit

**Comment by blever** [comment]
@FrontierOtr good to merge?

**Pull request merged**